Tracking doc overhaul

B-Step62 opened the pull request:
### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

This PR refactors [MLflow Tracking](https://mlflow.org/docs/latest/tracking.html) page to be consice and user-friendly.

**Preface**
Creating a huge PR includes all changes while knowing it could be impossible to review at once, but just to provide the whole preview and gather overalll feedbacks. Maybe will split to smaller steps once the high-level structure is fixed. For the time being, please focus on contents and structure, not detailed wording / formatting.

**Main changes**
- Splitted the whole page with > 1k lines and extract details to subpages.
- Sorted contents to align with users' onboarding journey.
- Refactored "Scenario" sections to present use cases and visuals, then hide implementation details.
- Added tutorials to set up common tracking environment.
- Added FAQ section to list common requests like "can we launch parallel runs?"

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

B-Step62 committed:
Start Tracking doc overhaul

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Further edit (mainly for client side)

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Add backend/artifacts store pages

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Add server page

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Add tracking set up tutorials

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Tweak scenario 2

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Fix and write actual tutorials

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Fix typo and lint

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

harupy commented on docs/source/tracking/index.rst:
```diff
@@ -0,0 +1,311 @@
+
+    tracking-api
+
+`MLflow Tracking APIs <tracking-api.html>`_ provide a set of functions to track your runs. The quickest way to start tracking is to use `Auto-logging <autolog.html>`, a powerful feature that allows you to log metrics, parameters, and models without the need for explicit log statements.
+All you need to do is call :py:func:`mlflow.autolog` before your training code.
+
+    .. code-block:: python
+
+        import mlflow
+
+        mlflow.autolog()
+
+        # Your training code...
```
IMO, autolog is an advanced feature. I think we can provide a primitive example here:

```python
with mlflow.start_run():
    mlflow.log_param("p", 1)
    mlflow.log_metric("m", 2)
```

B-Step62 replied:
I see, tried to sell that convenience of autolog... but yea I can put that example as well on the top as a starter:)

harupy commented:
@B-Step62 Could you preview the tracking page on the mobile view?

<img width="804" alt="image" src="https://github.com/mlflow/mlflow/assets/17039389/e39d60b7-a2fa-443c-bbc2-11165d805638">

BenWilson2 commented on docs/source/index.rst:
```diff
@@ -273,7 +273,7 @@ Deployment Guides and Tutorials
     traditional-ml/index
     deployment/index
-    tracking
+    tracking/index
```
We'll want to create a stub redirect page similar to what was done for `registry` / `model-registry` in order to not break external links to `tracking#<x>` (I'm sure there are links in Databricks docs that point to stuff in this page). Can you take a look at how the registry redirect was done and ensure that we're doing a 'hidden' page (not listed in ToC) that preserves the old name that will redirect to the content of the refactored main page (index.html)?

B-Step62 replied:
Totally makess sense, will follow the same way as registry page

B-Step62 replied:
@BenWilson2 It seems the approach of [registry](https://github.com/mlflow/mlflow/blob/master/docs/source/registry.rst) page doesn't work here as the old/new docs are in different directory. Like if we do this:

```
:orphan:
.. include:: tracking/index.rst
  :start-line: 1
```

it simply copies the scripts in `tracking/index.rst` to `tracking.rst`, then all relative path references will break e.g. static resources.

I also tried [sphinxcontrib-redirects](https://pypi.org/project/sphinxcontrib-redirects/) plugin, but unfortunately it only supports page level redirection and doesn't keep section `tracking#<x>`.

We might be able to implement custom JS solution, but since we will be migrating to markdown-based framework, I'm afraid it will be a thrown-away anyway. Hence, in this PR I will just keep `tracking.rst` file instead of `tracking/index.rst`, and just include child pages in `tracking/` directory.

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+
+The artifact store is a core component in `MLflow Tracking <../index.html>`_ where MLflow stores (typicaly large) arifacts
+for each run such as model weights (e.g. a pickled scikit-learn model), images (e.g. PNGs), model and data files (e.g. `Parquet <https://parquet.apache.org/>`_ file). 
+Note that metadata like parameters, metrics, and tags are stored in `backend store <backend-stores.html>`_, the other component of the MLflow Tracking.
```
```suggestion
Note that metadata like parameters, metrics, and tags are stored in a `backend store <backend-stores.html>`_ (e.g., PostGres, MySQL, or MSSQL Database), the other component of the MLflow Tracking.
```
I think listing out examples of the backend stores might be helpful, just like you did with the artifact store examples :)

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+for each run such as model weights (e.g. a pickled scikit-learn model), images (e.g. PNGs), model and data files (e.g. `Parquet <https://parquet.apache.org/>`_ file). 
+Note that metadata like parameters, metrics, and tags are stored in `backend store <backend-stores.html>`_, the other component of the MLflow Tracking.
+
+Configure Artifact Store
```
```suggestion
Configuring an Artifact Store
```
I think going with the gerund form fits in nicely with your other h2 headings below, as well as the instructional style of content

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+Configure Artifact Store
+========================
+MLflow by default stores artifacts in local ``./mlruns`` directory, but also supports various locations suitable for large data:
+AAmazon S3, Azure Blob Storage, Google Cloud Storage, SFTP server, and NFS. You can connect those remote storages via MLflow Tracking server.
```
```suggestion
Amazon S3, Azure Blob Storage, Google Cloud Storage, SFTP server, and NFS. You can connect those remote storages via MLflow Tracking server.
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+========================
+MLflow by default stores artifacts in local ``./mlruns`` directory, but also supports various locations suitable for large data:
+AAmazon S3, Azure Blob Storage, Google Cloud Storage, SFTP server, and NFS. You can connect those remote storages via MLflow Tracking server.
+See :ref:`tracking server setup <tracking-server-artifact-store>` and the specific section for your storage in :ref:`supported storages <artifacts-store-supported-storages>` for 
+how to connect to the remote storage.
```
```suggestion
See :ref:`tracking server setup <tracking-server-artifact-store>` and the specific section for your storage in :ref:`supported storages <artifacts-store-supported-storages>` for guidance on 
how to connect to your remote storage of choice.
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+
+.. _artifacts-stores-manage-access:
+
+Manage Access
```
```suggestion
Managing Artifact Store Access
```

Gerund form reads more like the h2 heading on line 35

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+Manage Access
+-------------
+To allow the server and clients to access the artifact location, you should configure your cloud
+provider credentials as normal. For example, for S3, you can set the ``AWS_ACCESS_KEY_ID``
```
```suggestion
provider credentials as you would for accessing them in any other capacity. For example, for S3, you can set the ``AWS_ACCESS_KEY_ID``
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+profile in ``~/.aws/credentials``.
+
+.. important::
+    Access credentials and configuration for the artifact storage location are configured *once during server initialization* in the place
+    of having users handle access credentials for artifact-based operations. Note that *all users who have access to the
+    Tracking Server in this mode will have access to artifacts served through this assumed role*.
```
```suggestion
    Access credentials and configuration for the artifact storage location are configured **once during server initialization** in the place
    of having users handle access credentials for artifact-based operations. Note that **all users who have access to the
    Tracking Server in this mode will have access to artifacts served through this assumed role**.
```
This is pretty important. Let's go with `<strong>` vs `<i>` here

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+    of having users handle access credentials for artifact-based operations. Note that *all users who have access to the
+    Tracking Server in this mode will have access to artifacts served through this assumed role*.
+
+Timeout
```
```suggestion
Setting an access Timeout
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+Timeout
+-------
+You can set an environment variable ``MLFLOW_ARTIFACT_UPLOAD_DOWNLOAD_TIMEOUT`` (in seconds) to configure the timeout for artifact uploads and downloads.
+Note that this is experimental feature, may be changed or removed.
```
```suggestion
Note that this is an experimental feature which may be changed or removed.
```
Can we indicate what the default value is for this environment variable so that users don't have to look it up in source code?

B-Step62 replied:
Makes sense! fyi apparently this one follows default timeout for each storage provider.

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+You can set an environment variable ``MLFLOW_ARTIFACT_UPLOAD_DOWNLOAD_TIMEOUT`` (in seconds) to configure the timeout for artifact uploads and downloads.
+Note that this is experimental feature, may be changed or removed.
+
+Logging Artifacts Location
```
```suggestion
Setting a Default Artifact Location for Logging
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+
+Logging Artifacts Location
+--------------------------
+MLflow automatically records ``artifact_uri`` property as a part of :py:class:`mlflow.entities.RunInfo`, so you can
```
```suggestion
MLflow automatically records the ``artifact_uri`` property as a part of :py:class:`mlflow.entities.RunInfo`, so you can
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+--------------------------
+MLflow automatically records ``artifact_uri`` property as a part of :py:class:`mlflow.entities.RunInfo`, so you can
+retrieve the location of the artifacts for historical runs using the :py:func:`mlflow.get_artifact_uri` API. 
+Also, ``artifact_location`` is a property recorded on :py:class:`mlflow.entities.Experiment` for
+default location to store artifacts for all runs in this experiment.
```
```suggestion
Also, ``artifact_location`` is a property recorded on :py:class:`mlflow.entities.Experiment` for setting the 
default location to store artifacts for all runs in a given experiment.
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+
+  If you do not specify a ``--default-artifact-root`` or an artifact URI when creating the experiment
+  (for example, ``mlflow experiments create --artifact-location s3://<my-bucket>``), the artifact root
+  is a path inside the file store. Typically this is not an appropriate location, as the client and
```
```suggestion
  will be set as a path inside the local file store (the hard drive of the computer executing your run). Typically this is not an appropriate location, as the client and
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+
+.. _artifacts-store-supported-storages:
+
+Supported storages
```
```suggestion
Supported storage types for the Artifact Store
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+these are available. For more information on how to set credentials, see
+`Set up AWS Credentials and Region for Development <https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/setup-credentials.html>`_.
+
+To add S3 file upload extra arguments, set ``MLFLOW_S3_UPLOAD_EXTRA_ARGS`` to a JSON object of key/value pairs.
```
For searchability, we might want to wrap each of these examples and their explanation introductions in an h3 header ``^^^^``. These examples are pretty important.

B-Step62 replied:
True, will make them distinct subsections.

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+        For example, providing ``--default-artifact-root $MLFLOW_S3_ENDPOINT_URL`` on the server side **and** ``MLFLOW_S3_ENDPOINT_URL`` on the client side will create a client path resolution issue for the artifact storage location.
+        Upon resolving the artifact storage location, the MLflow client will use the value provided by ``--default-artifact-root`` and suffixes the location with the values provided in the environment variable  ``MLFLOW_S3_ENDPOINT_URL``.
+        Depending on the value set for the environment variable ``MLFLOW_S3_ENDPOINT_URL``, the resulting artifact storage path for this scenario would be one of the following invalid object store paths:  ``https://<bucketname>.s3.<region>.amazonaws.com/<key>/<bucketname>/<key>`` or  ``s3://<bucketname>/<key>/<bucketname>/<key>``.
+        To prevent path parsing issues, ensure that reserved environment variables are removed (``unset``) from client environments.
```
```suggestion
        To prevent path parsing issues, **ensure that reserved environment variables are removed (``unset``) from client environments**.
```
emphasis

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+        Depending on the value set for the environment variable ``MLFLOW_S3_ENDPOINT_URL``, the resulting artifact storage path for this scenario would be one of the following invalid object store paths:  ``https://<bucketname>.s3.<region>.amazonaws.com/<key>/<bucketname>/<key>`` or  ``s3://<bucketname>/<key>/<bucketname>/<key>``.
+        To prevent path parsing issues, ensure that reserved environment variables are removed (``unset``) from client environments.
+
+Complete list of configurable values for an S3 client is available in `boto3 documentation <https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#configuration>`_.
```
```suggestion
The complete list of configurable parameters for an S3 client is available in the `boto3 documentation <https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#configuration>`_.
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+
+To store artifacts in Azure Blob Storage, specify a URI of the form
+``wasbs://<container>@<storage-account>.blob.core.windows.net/<path>``.
+MLflow expects Azure Storage access credentials in the
```
```suggestion
MLflow expects that your Azure Storage access credentials are located in the
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+To store artifacts in Azure Blob Storage, specify a URI of the form
+``wasbs://<container>@<storage-account>.blob.core.windows.net/<path>``.
+MLflow expects Azure Storage access credentials in the
+``AZURE_STORAGE_CONNECTION_STRING``, ``AZURE_STORAGE_ACCESS_KEY`` environment variables
```
```suggestion
``AZURE_STORAGE_CONNECTION_STRING`` and ``AZURE_STORAGE_ACCESS_KEY`` environment variables
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+#. ``AZURE_STORAGE_ACCESS_KEY``
+#. ``DefaultAzureCredential()``
+
+You must set one of these options on both your client application and your MLflow tracking server.
```
```suggestion
You must set one of these options on **both your client application and your MLflow tracking server**.
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+There are also two ways to authenticate to HDFS:
+
+- Use current UNIX account authorization
+- Kerberos credentials using following environment variables:
```
```suggestion
- Kerberos credentials using the following environment variables:
```

BenWilson2 commented on docs/source/tracking/artifacts-stores.rst:
```diff
@@ -0,0 +1,200 @@
+Most of the cluster contest settings are read from ``hdfs-site.xml`` accessed by the HDFS native
+driver using the ``CLASSPATH`` environment variable.
+
+The used HDFS driver is ``libhdfs``.
```
```suggestion
The HDFS driver that is used is ``libhdfs``.
```

BenWilson2 commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,353 @@
+
+.. note::
+
+    Generic autolog function :py:func:`mlflow.autolog` enables autologging for each supported library you have installed as soon as you import it.
```
```suggestion
    The generic autolog function :py:func:`mlflow.autolog` enables autologging for each supported library you have installed as soon as you import it.
```

BenWilson2 commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,353 @@
+.. note::
+
+    Generic autolog function :py:func:`mlflow.autolog` enables autologging for each supported library you have installed as soon as you import it.
+    Alternatively, you can use library-specific autolog calls such as :py:func:`mlflow.pytorch.autolog` to explicitly enable autologging for a particular library.
```
enable (or disable)

BenWilson2 commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,353 @@
+    )
+
+See :py:func:`mlflow.autolog` for the full set of arguments you can use.
+
```
Can we add a section with an example for disabling a particular flavor's autolog functionality (we don't need a full example in code for why this would be needed, but we can mention that if someone is using sklearn for building a pre-processing feature engineering step, they might not want that model to be automatically logged; but rather, they want the torch model that consumes that data to be autologged)

B-Step62 replied:
Ah that is an interesting but common use case, will add that example, thanks!

BenWilson2 commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,353 @@
+-----
+Call the generic autolog function or :py:func:`mlflow.tensorflow.autolog` before your training code to enable automatic logging of metrics and parameters. As an example, try running the `Keras/Tensorflow example <https://github.com/mlflow/mlflow/blob/master/examples/keras/train.py>`_.
+
+Note that only ``tensorflow>=2.3`` are supported.
```
```suggestion
Note that only versions of ``tensorflow>=2.3`` are supported.
```

BenWilson2 commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,353 @@
+If no active run exists when ``autolog()`` captures data, MLflow will automatically create a run to log information to.
+Also, MLflow will then automatically end the run once training ends via calls to ``tf.keras.fit()``.
+
+If a run already exists when ``autolog()`` captures data, MLflow will log to that run but not automatically end that run after training.
```
"you will have to manually stop the run if you wish to start a new run context for logging to a new run." or something to that effect.

BenWilson2 commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,353 @@
+
+.. _autolog-xgboost:
+
+XGBoost
```
nit: 
Can we either:
1. Sort these alphabetically
2. Sort these according to general popularity? (torch, XGBoost, TF / Keras, sklearn, lightgbm, .... <way at the end: Gluon>)

B-Step62 replied:
Good idea! I'll go with 1, as it's easier for ppl to assume the order, also not sure if the popularity thing remain stable in the future:p

BenWilson2 commented on docs/source/tracking/backend-stores.rst:
```diff
@@ -0,0 +1,105 @@
+* Code version (only if you launch runs from an :ref:`MLflow Project <projects>`).
+* Source file name (only if you launch runs from an :ref:`MLflow Project <projects>`).
+
+Note that large model artifacts such as model weight files are stored in `artifact store <artifacts-stores.html>`_,  the other core component in MLflow Tracking.
```
```suggestion
Note that large model artifacts such as model weight files are stored in `artifact store <artifacts-stores.html>`_,  the other core components are stored in your backend store (MLflow Tracking).
```

B-Step62 replied:
oops I meant sth like "artifact store, which is the other component in MLflow tracking", will rephase a bit

BenWilson2 commented on docs/source/tracking/backend-stores.rst:
```diff
@@ -0,0 +1,105 @@
+Configure Backend Store
+=======================
+By default, MLflow stores metadata in local files in the ``./mlruns`` directory, but MLflow can store metadata to databases as well.
+You can configure the location by passing **tracking URI** to MLflow, via either of the following methods:
```
```suggestion
You can configure the location by passing the desired **tracking URI** to MLflow, via either of the following methods:
```

BenWilson2 commented on docs/source/tracking/backend-stores.rst:
```diff
@@ -0,0 +1,105 @@
+
+* Set the ``MLFLOW_TRACKING_URI`` environment variable.
+* Call :py:func:`mlflow.set_tracking_uri` in your code.
+* If you are running :ref:`Tracking Server <tracking_server>`, you can set the ``tracking_uri`` option when starting the server, like ``mlflow server --backend-store-uri sqlite:///mydb.sqlite``
```
```suggestion
* If you are running a :ref:`Tracking Server <tracking_server>`, you can set the ``tracking_uri`` option when starting the server, like ``mlflow server --backend-store-uri sqlite:///mydb.sqlite``
```

BenWilson2 commented on docs/source/tracking/backend-stores.rst:
```diff
@@ -0,0 +1,105 @@
+=====================
+MLflow supports the following types of tracking URI for backend stores:
+
+- Local file path (specified as ``file:/my/local/dir``), where data is just directly stored locally.
```
```suggestion
- Local file path (specified as ``file:/my/local/dir``), where data is just directly stored locally to a system disk where your code is executing.
```

BenWilson2 commented on docs/source/tracking/backend-stores.rst:
```diff
@@ -0,0 +1,105 @@
+MLflow supports the following types of tracking URI for backend stores:
+
+- Local file path (specified as ``file:/my/local/dir``), where data is just directly stored locally.
+- Database encoded as ``<dialect>+<driver>://<username>:<password>@<host>:<port>/<database>``. MLflow supports the dialects ``mysql``, ``mssql``, ``sqlite``, and ``postgresql``. For more details, see `SQLAlchemy database uri <https://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls>`_.
```
```suggestion
- A Database, encoded as ``<dialect>+<driver>://<username>:<password>@<host>:<port>/<database>``. MLflow supports the dialects ``mysql``, ``mssql``, ``sqlite``, and ``postgresql``. For more details, see `SQLAlchemy database uri <https://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls>`_.
```

BenWilson2 commented on docs/source/tracking/backend-stores.rst:
```diff
@@ -0,0 +1,105 @@
+  Refer to Access the MLflow tracking server from outside Databricks `[AWS] <http://docs.databricks.com/applications/mlflow/access-hosted-tracking-server.html>`_
+  `[Azure] <http://docs.microsoft.com/azure/databricks/applications/mlflow/access-hosted-tracking-server>`_, or :ref:`the quickstart <quickstart_tracking_server>` to
+  easily get started with hosted MLflow on Databricks Community Edition.
+
```
AzureML?

B-Step62 replied:
Hmm as far as I read the linked doc, there is no mention about AzureML... It describes how to use MLflow outside AD

BenWilson2 commented on docs/source/tracking/backend-stores.rst:
```diff
@@ -0,0 +1,105 @@
+  `[Azure] <http://docs.microsoft.com/azure/databricks/applications/mlflow/access-hosted-tracking-server>`_, or :ref:`the quickstart <quickstart_tracking_server>` to
+  easily get started with hosted MLflow on Databricks Community Edition.
+
+.. note::
```
We should change this to "important". So many people ask this question. Let's emphasize text to make it stand out as well.

B-Step62 replied:
Got it! Also worth add it to the FAQ section then.

BenWilson2 commented on docs/source/tracking/backend-stores.rst:
```diff
@@ -0,0 +1,105 @@
+======================
+
+MLflow will automatically try to use `LibYAML <https://pyyaml.org/wiki/LibYAML>`_ bindings if they are already installed.
+However if you notice any performance issues when using *file store* backend, it could mean LibYAML is not installed on your system.
```
```suggestion
However, if you notice any performance issues when using *file store* backend, it could mean LibYAML is not installed on your system.
```

BenWilson2 commented on docs/source/tracking/index.rst:
```diff
@@ -0,0 +1,311 @@
+
+        # Your training code...
+
+Auto-logging supports popular libraries such as :ref:`Scikit-learn <autolog-sklearn>`, :ref:`XGBoost <autolog-xgboost>`, :ref:`PyTorch <autolog-pytorch>`, :ref:`Keras <autolog-keras>`, :ref:`Spark <autolog-spark>`, and more.
```
Can we convert this to a bulleted list so that the links are more distinct?

B-Step62 replied:
Yes I thought of that, but one downside is that it took more space and makes the page a bit sparse. Given this is a gateway page, I prioritized consiceness here, WDYT?

BenWilson2 commented on docs/source/tracking/index.rst:
```diff
@@ -0,0 +1,311 @@
+------------------------------
+
+You can also access all of the functions in the Tracking UI programmatically with :py:class:`MlflowClient <mlflow.client.MlflowClient>`.
+
```
Should we add a quick search runs example here?

BenWilson2 commented on docs/source/tracking/index.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. _tracking-setup:
+
+Set up MLflow Tracking Environment
```
```suggestion
Set up the MLflow Tracking Environment
```

BenWilson2 commented on docs/source/tracking/index.rst:
```diff
@@ -0,0 +1,311 @@
+    * - **2. Local Tracking with Local Database**
+      - Solo development
+      - The MLflow client can interface with a SQLAlchemy-compatible database (e.g., SQLite, PostgreSQL, MySQL) for the `backend <backend-stores.html>`_. Saving metadata to a database allows you cleaner management of your experiment data while skipping the effort of setting up a server.
+      - `Tracking Experiments with Local Database <tutorials/local-database.html>`_
```
```suggestion
      - `Tracking Experiments using a Local Database <tutorials/local-database.html>`_
```

BenWilson2 commented on docs/source/tracking/index.rst:
```diff
@@ -0,0 +1,311 @@
+
+.. |image-artifact-only| image:: ../_static/images/tracking/tracking-setup-artifacts-only.png
+    :align: middle
+    :width: 50px
```
Can we set % of DOM display element instead of fixed size just to make sure that screen resizing scales correctly?

B-Step62 replied:
Absolutely, should've replaced the test size.

BenWilson2 commented on docs/source/tracking/index.rst:
```diff
@@ -0,0 +1,311 @@
+
+* :ref:`Organize runs into experiments <organizing_runs_in_experiments>` - Experiments are logical containers for your runs. You can create an experiment using the CLI, API, or UI.
+* :ref:`Add tags to runs <add-tags-to-runs>` - You can associate arbitrary tags with each run, which allows you to filter and search runs based on tags.
+
```
child runs too! :)

B-Step62 replied:
Good point! Will add it.

BenWilson2 commented on docs/source/tracking/server.rst:
```diff
@@ -0,0 +1,247 @@
+.. important:: 
+  The server listens on http://localhost:5000 by default and only accepts
+  connections from the local machine. To let the server accept connections
+  from other machines, you will need to pass ``--host 0.0.0.0`` to listen on
```
Can we mention that this is a required conf for deployment inside k8s / docker containers (generally)?

B-Step62 replied:
Sure, adding real use case should help users (+ searchability)

BenWilson2 commented on docs/source/tracking/server.rst:
```diff
@@ -0,0 +1,247 @@
+        remote_server_uri = "..."  # set to your server URI
+        mlflow.set_tracking_uri(remote_server_uri)
+        # Note: on Databricks, the experiment name passed to mlflow_set_experiment must be a
+        # valid path in the workspace
```
Can we add a fake User/<email>/path example so that Databricks Users will instantly recognize what this is referring to?

BenWilson2 commented on docs/source/tracking/server.rst:
```diff
@@ -0,0 +1,247 @@
+            mlflow.log_param("a", 1)
+            mlflow.log_metric("b", 2)
+
+    .. code-block:: R
```
Java (or Scala, even better) example as well? :)

B-Step62 replied:
Yup I can add it :)

BenWilson2 commented on docs/source/tracking/server.rst:
```diff
@@ -0,0 +1,247 @@
+This will create a SQLite database ``my.db`` in the current directory, and logging requests from clients will be pointed to this database.
+
+.. note::
+  If you are using SQL-based backend store, artifact store URI has to be specified otherwise the server will fail to start (because we can't store artifacts to SQL database).
```
```suggestion
  If you are using a SQL-based backend store, the artifact store URI has to be specified, otherwise the server will fail to start (because we can't store artifacts to a SQL database).
```

BenWilson2 commented on docs/source/tracking/server.rst:
```diff
@@ -0,0 +1,247 @@
+
+.. note::
+  If you are using SQL-based backend store, artifact store URI has to be specified otherwise the server will fail to start (because we can't store artifacts to SQL database).
+  Refer to the section below to how to configure it.
```
```suggestion
  Refer to the section below for details on how to configure the artifact store URI when using a SQL-based backend store.
```

BenWilson2 commented on docs/source/tracking/tracking-api.rst:
```diff
@@ -0,0 +1,281 @@
+The run remains open throughout the ``with`` statement, and is automatically closed when the
+statement exits, even if it exits due to an exception.
+
+
```
Can we show a parent + multiple child runs as an example and explain that it's common to use this setup for hyperparameter tuning?

B-Step62 replied:
Absolutely! Indeed I feel that might be the main ask of the said question.

BenWilson2 commented on docs/source/tracking/tracking-api.rst:
```diff
@@ -0,0 +1,281 @@
++-------------------------------+----------------------------------------------------------------------------------------+
+| Key                           | Description                                                                            |
++===============================+========================================================================================+
+| ``mlflow.note.content``       | A descriptive note about this run. This reserved tag is not set automatically and can  |
```
Can we call out in the text above the table that this is an Exception and is the best way to programmatically pre-seed the notes location? A LOT of new users ask about this.

B-Step62 replied:
Ooh I see, yea definitely otherwise it's confusing to use reserved key. Also worth having subsection and FAQ imo.

BenWilson2 commented on docs/source/tracking/tracking-api.rst:
```diff
@@ -0,0 +1,281 @@
+Get MLflow Run instance from autologged results
+-----------------------------------------------
+
+In some cases, you may want to access the MLflow Run instance associated with the autologged results, similarly to you can get `mlflow.start_run()` 
```
```suggestion
In some cases, you may want to access the MLflow Run instance associated with the autologged results, similarly to you can get the run context with `mlflow.start_run()` 
```

BenWilson2 commented on docs/source/tracking/tracking-api.rst:
```diff
@@ -0,0 +1,281 @@
+
+    # Use the model to make predictions on the test dataset.
+    predictions = rf.predict(X_test)
+    autolog_run = mlflow.last_active_run()
```
Can we show an example of what the return type is and show the stdout return of `mlflow.last_active_run()`?

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+========================================
+Tracking Experiments with Local Database
```
```suggestion
Tracking Experiments with a Local Database
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+Tracking Experiments with Local Database
+========================================
+
+In this tutorial, you will learn how to use local database to track your experiment metadata with MLflow. By default, MLflow Tracking logs run to local files,
```
```suggestion
In this tutorial, you will learn how to use a local database to track your experiment metadata with MLflow. By default, MLflow Tracking logs run data to local files,
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+========================================
+
+In this tutorial, you will learn how to use local database to track your experiment metadata with MLflow. By default, MLflow Tracking logs run to local files,
+which may cause some frustration due to fractured small files and lack of simple access interface. Also, if you are using Python, you can use SQLite that runs 
```
```suggestion
which may cause some frustration due to fractured small files and the lack of a simple access interface. Also, if you are using Python, you can use SQLite that runs 
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+
+In this tutorial, you will learn how to use local database to track your experiment metadata with MLflow. By default, MLflow Tracking logs run to local files,
+which may cause some frustration due to fractured small files and lack of simple access interface. Also, if you are using Python, you can use SQLite that runs 
+upon local file (e.g. ``mlruns.db``) and has a built-in client ``sqlite3``, eliminating the effort to install any additional dependencies and setting up database server.
```
```suggestion
upon your local file system (e.g. ``mlruns.db``) and has a built-in client ``sqlite3``, eliminating the effort to install any additional dependencies and setting up database server.
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+Step 2 - Configure MLflow environment varialbles
+------------------------------------------------
+
+Set tracking URI to local SQLite database
```
```suggestion
Set the tracking URI to a local SQLite database
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+Set tracking URI to local SQLite database
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To point MLflow to your local SQLite database, you need to set envioronment variable ``MLFLOW_TRACKING_URI`` to ``sqlite:///mlruns.db``.
```
```suggestion
To point MLflow to your local SQLite database, you need to set the environment variable ``MLFLOW_TRACKING_URI`` to ``sqlite:///mlruns.db``.
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To point MLflow to your local SQLite database, you need to set envioronment variable ``MLFLOW_TRACKING_URI`` to ``sqlite:///mlruns.db``.
+(This will create a SQLite database file called ``mlruns.db`` in the current directory. Specify different path if you want to store the database file in a different location.)
```
```suggestion
(This will create a SQLite database file called ``mlruns.db`` in the current directory. Specify a different path if you want to store the database file in a different location.)
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+        %env MLFLOW_TRACKING_URI=sqlite:///mlruns.db
+
+.. note::
+  For SQLite database, MLflow automatically creates a new database if it does not exist. If you want to use other database, you need to create the database first.
```
```suggestion
  For using a SQLite database, MLflow automatically creates a new database if it does not exist. If you want to use another database, you need to create the database first.
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+Step 3: Start logging
+---------------------
+
+Now you are ready to start logging your experiment runs. For example, the following code runs training for an scikit-learn RandomForest model on diabetes dataset:
```
```suggestion
Now you are ready to start logging your experiment runs. For example, the following code runs training for an scikit-learn RandomForest model on diabetes dataset:
```
```suggestion
Now you are ready to start logging your experiment runs. For example, the following code runs training for an scikit-learn RandomForest model on the diabetes dataset:
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+        # Use the model to make predictions on the test dataset.
+        predictions = rf.predict(X_test)
+
+Step 4: View logged Run in Tracking UI
```
```suggestion
Step 4: View your logged Run in the Tracking UI
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+Step 4: View logged Run in Tracking UI
+--------------------------------------
+
+Once your training job finishes, you can run following command to launch the MLflow UI. You have to specify the path to SQLite database file with ``--backend-store-uri`` option.
```
```suggestion
Once your training job finishes, you can run the following command to launch the MLflow UI (You will have to specify the path to SQLite database file with ``--backend-store-uri`` option): 
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+What's Next?
+============
+
+Now you have learned how to connect MLflow Tracking with remote storage and database.
```
```suggestion
You've now learned how to connect MLflow Tracking with a remote storage and a database.
```

BenWilson2 commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+There are a couple of more advanced topics you can explore:
+
+* **Remote environment setup for team development**: While storing runs and experiments data in local machine is perfectly fine for solo development, you should 
+  consider using :ref:`MLflow Tracking Server <tracking_server>` when you set up team collaboration environment with MLflow Tracking. Read 
```
```suggestion
  consider using :ref:`MLflow Tracking Server <tracking_server>` when you set up a team collaboration environment with MLflow Tracking. Read 
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+Remote Experiment Tracking with MLflow Tracking Server
+======================================================
+
+In this tutorial, you will learn how to set up MLflow Tracking environment for team development using :ref:`MLflow Tracking Server <tracking_server>`.
```
```suggestion
In this tutorial, you will learn how to set up MLflow Tracking environment for team development using the :ref:`MLflow Tracking Server <tracking_server>`.
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+There are many benefits to utilize MLflow Tracking Server for remote experiment tracking:
+
+* **Collaboration**: Multiple users can log runs to the same endpoint, and query runs and models logged by other users.
+* **Sharing Results**: The tracking server also serves :ref:`tracking_ui` endpoint, where team members can easily explore each other's results.
```
```suggestion
* **Sharing Results**: The tracking server also serves a :ref:`tracking_ui` endpoint, where team members can easily explore each other's results.
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+* **Sharing Results**: The tracking server also serves :ref:`tracking_ui` endpoint, where team members can easily explore each other's results.
+* **Centralized Access**: The tracking server can be run as a proxy for the remote access for metadata and artifacts, making it easier to secure and audit access to data.
+
+How it works?
```
```suggestion
How does it work?
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+How it works?
+=============
+
+The following picture depicts the architecture of using remote MLflow Tracking Server with PostgreSQL and S3
```
```suggestion
The following picture depicts the architecture of using a remote MLflow Tracking Server with PostgreSQL and S3
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+    Artifacture diagram of MLflow Tracking Server with PostgreSQL and S3
+
+.. note::
+   You cna find the list of supported data stores in `artifact stores <../artifacts-stores.html>`_ and `backend stores <../backend-stores.html>` documentations.
```
```suggestion
   You can find the list of supported data stores in the `artifact stores <../artifacts-stores.html>`_ and `backend stores <../backend-stores.html>` documentation guides.
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+  * Artifacts are passed to the end user through the Tracking Server through the interface of the ``HttpArtifactRepository``
+
+
+Get Started
```
```suggestion
Getting Started
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+Preface
+-------
+
+In the actual environment, you will have a multiple remote hosts to run the tracking server and databases, as shown in the diagram above. However, for the pupose of this tutorial, 
```
```suggestion
In an actual production deployment environment, you will have multiple remote hosts to run both the tracking server and databases, as shown in the diagram above. However, for the purposes of this tutorial, 
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+-------
+
+In the actual environment, you will have a multiple remote hosts to run the tracking server and databases, as shown in the diagram above. However, for the pupose of this tutorial, 
+we will just use single machine with multiple Docker containers running on different ports, mimicing the remote environment with easy set up. We also use `MinIO <https://min.io/>`_,
```
```suggestion
we will just use a single machine with multiple Docker containers running on different ports, mimicking the remote environment with a far easier evaluation tutorial setup. We will also use `MinIO <https://min.io/>`_,
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+Step 1 - Get MLflow and additional dependencies
+-----------------------------------------------
+MLflow is available on PyPI. Also `pyscopg2 <https://pypi.org/project/psycopg2/>`_ and `boto3 <https://boto3.amazonaws.com/v1/documentation/api/latest/index.html>`_ are required for accessing PostgreSQL and S3 with Python.
+If you don't already have them installed on your system, you can install it with:
```
```suggestion
If you don't already have them installed on your system, you can install them with:
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+Step 2 - Set up remote data stores
+----------------------------------
+MLflow Tracking Server can interact with a variety of data stores to store experiment and run data as well as artifacts.
+In this tutorial, we will use **Docker Compose** to start two containers, each of them simulating remote servers in actual environment.
```
```suggestion
In this tutorial, we will use **Docker Compose** to start two containers, each of them simulating remote servers in an actual environment.
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+----------------------------------
+
+.. note::
+  In actual environment, you will have remote host to run the tracking server, but in this tutorial we will just use the local machine.
```
```suggestion
  In actual environment, you will have a remote host that will run the tracking server, but in this tutorial we will just use our local machine as a simulated surrogate for a remote machine.
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+
+You can find the instructions for how to configure credentials for other storages in :ref:`Supported Storage <artifacts-store-supported-storages>`.
+
+Launch tracking server
```
```suggestion
Launch the tracking server
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+Step 4: Logging to the Tracking Server
+--------------------------------------
+
+Once the tracking server is running, you can log runs to the tracking server by setting the MLflow Tracking URI to the tracking server's URI. Alternatively, you can use the :py:func:`mlflow.set_tracking_uri` API to set the tracking URI.
```
```suggestion
Once the tracking server is running, you can log runs to it by setting the MLflow Tracking URI to the tracking server's URI. Alternatively, you can use the :py:func:`mlflow.set_tracking_uri` API to set the tracking URI.
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+
+.. code-block:: bash
+
+    export MLFLOW_TRACKING_URI=http://127.0.0.1:5000  # Replace with remote host name or IP address in actual environment
```
```suggestion
    export MLFLOW_TRACKING_URI=http://127.0.0.1:5000  # Replace with remote host name or IP address in an actual environment
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+
+    export MLFLOW_TRACKING_URI=http://127.0.0.1:5000  # Replace with remote host name or IP address in actual environment
+
+Then run your code with MLflow tracking APIs as usual,  the following code runs training for an scikit-learn RandomForest model on diabetes dataset:
```
```suggestion
Thenm run your code with MLflow tracking APIs as usual. The following code runs training for an scikit-learn RandomForest model on the diabetes dataset:
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+Step 5: View logged Run in Tracking UI
+--------------------------------------
+
+MLflow Tracking Server also hosts Tracking UI on the same endpoint.
```
```suggestion
Our pseudo-remote MLflow Tracking Server also hosts the Tracking UI on the same endpoint. In an actual deployment environment with a remote tracking server, this is also the case. 
```

BenWilson2 commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+* **Secure the Tracking Server**: The ``--host`` option exposes the service on all interfaces. If running a server in production, we
+  would recommend not exposing the built-in server broadly (as it is unauthenticated and unencrypted). Read :ref:`Secure Tracking Server <tracking-auth>`
+  for the best practices to secure the Tracking Server in production.
+* **New Features**: MLflow team constantly develops new features to support broader use cases. See `New Features <../../new-features/index.html>`_ to catch up with the latest features.
```
```suggestion
* **New Features**: The MLflow team and a host of community contributors constantly develops new features to support broader use cases. See `New Features <../../new-features/index.html>`_ to catch up with the latest features!
```

B-Step62 replied:
oh this is important:)

BenWilson2 commented:
This is AWESOME! A few nits and typos, some suggestions for commonly asked about / confusing parts for users that would be good to highlight. Other than those, this is looking fantastic! Great work @B-Step62 !!!

B-Step62 committed:
Apply suggestions from code review

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Yuki Watanabe <31463517+B-Step62@users.noreply.github.com>

B-Step62 commented:
> @B-Step62 Could you preview the tracking page on the mobile view?
> 
> <img alt="image" width="804" src="https://user-images.githubusercontent.com/17039389/284890270-e39d60b7-a2fa-443c-bbc2-11165d805638.png">
> It doesn't have to look perfect on mobile devices.

Ooh great callout...! pictures can be fixed easily I guess, not sure what to do for tables... will try a few ways to make it at least readable on mobile.

B-Step62 committed:
Address more comments

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Add sphinx-tabs and formatting layout

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Recover tracking.rst to avoid need for redirection

ref: https://github.com/mlflow/mlflow/pull/10471/files#r1408950671

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Add tabs.css from sphinx-tabs

Somehow sphinx-tabs extension doesn't load their CSS. Adding the css file manually as a dirty workaround.

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Fix links

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented:
Note: Table and tabs in this PR are not shown nicely on mobile device. Will be fixed in the follow-up PR like this: https://github.com/B-Step62/mlflow/pull/1

B-Step62 committed:
lint

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

harupy commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+
+    export MLFLOW_TRACKING_URI=http://127.0.0.1:5000  # Replace with remote host name or IP address in an actual environment
+
+Thenm run your code with MLflow tracking APIs as usual. The following code runs training for an scikit-learn RandomForest model on the diabetes dataset:
```
```suggestion
Then run your code with MLflow tracking APIs as usual. The following code runs training for an scikit-learn RandomForest model on the diabetes dataset:
```

harupy commented on docs/source/tracking/tutorials/local-database.rst:
```diff
@@ -0,0 +1,103 @@
+Step 3: Start logging
+---------------------
+
+Now you are ready to start logging your experiment runs. For example, the following code runs training for an scikit-learn RandomForest model on the diabetes dataset:
```
```suggestion
Now you are ready to start logging your experiment runs. For example, the following code runs training for a scikit-learn RandomForest model on the diabetes dataset:
```

harupy commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+Create ``docker-compose.yaml``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Create a file named ``docker-compose.yaml`` with the following content:
```
```suggestion
Create a file named ``compose.yaml`` with the following content:
```

Can we use `compose.yaml`?

https://docs.docker.com/compose/compose-file/03-compose-file/

harupy commented on docs/source/tracking/tutorials/remote-server.rst:
```diff
@@ -0,0 +1,260 @@
+          local_path = mlflow.artifacts.download_artifacts(mlflow_artifact_uri)
+
+          # Load the model
+          model = mlflow.sklearn.load_model(local_path)
```
Can we load a model using `runs:/{run_id}`?

B-Step62 replied:
oh didn't know that URI is available, not listed in the original doc haha

harupy commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,367 @@
+    X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)
+
+    rf = RandomForestRegressor(n_estimators=100, max_depth=6, max_features=3)
+    rf.fit(X_train, y_train)
```
```suggestion
    rf.fit(X_train, y_train)  # ...
```

can we add a comment on this line to clarify where logging is performed?

harupy commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,367 @@
+    rf = RandomForestRegressor(n_estimators=100, max_depth=6, max_features=3)
+    rf.fit(X_train, y_train)
+
+Step 3 - Execute Your Code
+--------------------------
+
+.. code-section::
+
+    .. code-block:: bash
+        :name: execute code
+
+        python YOUR_ML_CODE.py
```
I think we can remove this section.

B-Step62 commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,367 @@
+
+    import mlflow
+
+    # Disable autologging for scikit-learn, but enable it for other libraries
+    mlflow.sklearn.autolog(disable=True)
+    mlflow.autolog()
```
This is an example for showing how we disable specific library.

harupy replied:
Ah you're right :) I didn't read this section carefully.

B-Step62 committed:
Address comments

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

harupy commented on docs/theme/mlflow/static/css/tabs.css:
Can we add a permalink to the source?

harupy commented on docs/source/tracking/autolog.rst:
```diff
@@ -0,0 +1,368 @@
+
+See :py:func:`mlflow.autolog` for the full set of arguments you can use.
+
+Disable Autologging for Specific Libraries
```
Can we add a "Enable Autologging for a specific library"?

B-Step62 replied:
Sure, added but consolidated into one section ("Enable / disable") to keep it concise.

BenWilson2 reviewed (approved):
Looks great! Let's merge :)
